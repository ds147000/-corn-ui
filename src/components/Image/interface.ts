import { ImageProps } from '@tarojs/components/types/Image'
import { formatImgOption } from 'src/utils/url'

export const DEFAULT_LIGHT_LAZY_IMG = 'data:image/jpg;base64,/9j/4AAQSkZJRgABAQAASABIAAD/2wCEAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAf/AABEIAGQAZAMBEQACEQEDEQH/xABNAAEBAAAAAAAAAAAAAAAAAAAAChABAAAAAAAAAAAAAAAAAAAAAAEBAQEAAAAAAAAAAAAAAAAAAAUGEQEAAAAAAAAAAAAAAAAAAAAA/9oADAMBAAIRAxEAPwC3BqE8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB/9k='
export const DEFAULT_LAZY_IMG = 'data:image/jpg;base64,/9j/4AAQSkZJRgABAQAASABIAAD/2wCEAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAf/AABEIAGQAZAMBEQACEQEDEQH/xABNAAEBAAAAAAAAAAAAAAAAAAAAChABAAAAAAAAAAAAAAAAAAAAAAEBAQEAAAAAAAAAAAAAAAAAAAQFEQEAAAAAAAAAAAAAAAAAAAAA/9oADAMBAAIRAxEAPwC6hsJwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAH/9k='



export interface XImageProps extends ImageProps, formatImgOption {
  /** 高亮风格 */
  light?: boolean
  /** 自定义懒加载占位图 */
  lazyImg?: string
  /** 放大预览 */
  previewImage?: boolean
}
