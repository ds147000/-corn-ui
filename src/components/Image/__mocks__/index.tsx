import Image from '../index.h5'

export default Image
