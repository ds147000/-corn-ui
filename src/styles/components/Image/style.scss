.xrk-img {
  width: unset;
  height: unset;
}
