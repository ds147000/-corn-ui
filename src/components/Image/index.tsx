/* eslint-disable import/export */
/* eslint-disable import/first */

// #if _APP === 'weapp'
import Image from './index.taro'

export default Image

// #else
import Image from './index.h5'


export default Image
// #endif
