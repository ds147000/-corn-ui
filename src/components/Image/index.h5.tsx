/* eslint-disable max-len */
import React, { useEffect, useMemo, useRef } from 'react'
import { Image } from '@tarojs/components'
import ClassNames from 'classnames'
import getObserverService from './utils/observer'
import { formatImg } from '../../utils/url'
import { DEFAULT_LAZY_IMG, DEFAULT_LIGHT_LAZY_IMG, XImageProps } from './interface'


const XImage: React.FC<XImageProps> = ({
  src,
  light,
  lazyImg,
  lazyLoad,
  m, w, h, l, s, limit, format,
  className,
  ...props
}) => {
  const el = useRef<{ imgRef: HTMLImageElement }>()
  const _class = useMemo(() => ClassNames('xrk-img', className), [ className ])

  // 缓冲的图片url
  const _src = useMemo(() => formatImg(src, { m, w, h, l, s, limit, format }), [ m, w, h, l, s, limit, format, src ])
  // 占位图
  const _lazyImg = light ? DEFAULT_LAZY_IMG : DEFAULT_LIGHT_LAZY_IMG

  useEffect(() => {
    if (!lazyLoad || !el.current) return

    const destory = getObserverService().add(el.current.imgRef, () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (el.current as any).imgRef.src = _src
    })

    // eslint-disable-next-line consistent-return
    return destory
  }, [ lazyLoad, _src ])

  return (
    <Image
      {...props}
      className={_class}
      lazyLoad={false}
      src={lazyLoad ? _lazyImg : _src}
      ref={el}
    />
  )
}

export default XImage
