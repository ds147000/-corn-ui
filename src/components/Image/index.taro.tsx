/* eslint-disable max-len */
import React, { useMemo } from 'react'
import { Image } from '@tarojs/components'
import { formatImg } from '../../utils/url'
import { XImageProps } from './interface'

const XImage: React.FC<XImageProps> = ({
  src,
  light,
  lazyImg,
  lazyLoad,
  m, w, h, l, s, limit, format,
  className,
  ...props
}) => {
  // 缓冲的图片url
  const _src = useMemo(() => formatImg(src, { m, w, h, l, s, limit, format }), [ m, w, h, l, s, limit, format, src ])

  return (
    <Image
      {...props}
      src={_src}
    />
  )
}

export default XImage
