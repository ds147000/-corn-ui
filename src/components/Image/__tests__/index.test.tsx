import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import Image from '../index.h5'

describe('Image', () => {
  const url = 'https://t7.baidu.com/it/u=1440354408,1578911780&fm=193&f=GIF'

  test('basis', async () => {
    const screen = render(<Image src={url} />)
    expect(screen.container.querySelector('img')?.src).toBe(url)
  })

  test('basis', async () => {
    const screen = render(<Image src={url} light />)
    expect(screen.container.querySelector('img')?.src).toBe(url)
    expect(screen.container).toMatchSnapshot()
  })

  test('lazyLoad', async () => {
    const screen = render(<Image src={url} lazyLoad />)
    expect(screen.container.querySelector('img')?.src).not.toBe(url)
  })

  test('lazyLoad Show', async () => {
    const screen = render(<Image src={url} lazyLoad />)
    expect(screen.container.querySelector('img')?.src).not.toBe(url)
    // eslint-disable-next-line no-magic-numbers
    await new Promise((res) => setTimeout(res, 1000))
    expect(screen.container.querySelector('img')?.src).toBe(url)
  })

  test('lazyLoad Show of destroy', async () => {
    const screen = render(<Image src={url} lazyLoad />)
    expect(screen.container.querySelector('img')?.src).not.toBe(url)
    screen.unmount()
    // eslint-disable-next-line no-magic-numbers
    await new Promise((res) => setTimeout(res, 1000))
    expect(screen.container.querySelector('img')).toBe(null)
  })

  test('preview', async () => {
    const screen = render(<Image src={url} previewImage />)
    fireEvent.click(screen.container.querySelector('img') as HTMLImageElement)
    await waitFor(() => expect(screen).toMatchSnapshot())
  })

  test('click', async () => {
    const click = jest.fn()
    const screen = render(<Image src={url} onClick={click} />)
    fireEvent.click(screen.container.querySelector('img') as HTMLImageElement)
    // eslint-disable-next-line no-magic-numbers
    expect(click).toHaveBeenCalledTimes(1)
  })
})
