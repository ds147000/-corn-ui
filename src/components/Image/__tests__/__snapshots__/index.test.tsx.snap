// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image basis 1`] = `
<div>
  <div
    class="taro-img xrk-img"
  >
    <img
      class="taro-img__mode-scaletofill"
      src="https://t7.baidu.com/it/u=1440354408,1578911780&fm=193&f=GIF"
    />
  </div>
</div>
`;
